### ISSUE TITLE
Editable timeout and new parameter

### ISSUE THREAD

**Issue by medvidekPanda** [capacitor-community/background-geolocation#34]
**Problem**

There is no option to set custom timeout and setFastestInterval parameter.

**Solution**

- Could you add possibility to add custom timeout value for setMaxWaitTime and setMaxWaitTime?
- Could you add setFastestInterval with possibility custom value?

**Comment by diachedelic** [comment]
This has been requested before, but I am reticent to include a configuration option for Android when an equivalent option does not exist for iOS. Is the distance filter option not sufficient for your  needs?

**Comment by medvidekPanda** [comment]
OK I didn't know that. But I need timeout, because I need request position periodically. For example every 30 seconds. I understand different options for different OS should be complicated. But in Cordova / Capacitor plugins it is normal. For my pour. iOS is not important for my purposes. I work on Android app only.

**Comment by diachedelic** [comment]
I am sympathetic to your cause. Could you please explain to me your use case, and why a distance filter is not sufficient?

**Comment by medvidekPanda** [comment]
Distance filter is not sufficient, because I need get location in long interval in some situation, for example every 30 seconds, but device is on one place. Yes I can make some workaround in app controler, but I think, get position from OS api is better, when API has this feature.

**Comment by diachedelic** [comment]
Yes, but why do you need the location every 30 seconds if the devices has not moved? Is it just to trigger something in the UI?

**Comment by medvidekPanda** [comment]
Exactly as you said. App runs on the foreground / background and every X seconds send location to server.

**Comment by diachedelic** [comment]
Why does the server require location updates every X seconds, even if the location has not changed?

**Comment by medvidekPanda** [comment]
It doesn't matter at all.  I have such a request from a customer.  He needs to be 100% sure that the device is in place or in motion at such a time and at such an interval.  When the object moves 10 km per hour, it travels a different distance than it travels 100 km per hour.

**Comment by diachedelic** [comment]
I see. The requests to the server are keeping the connection alive, so to speak.

**Comment by diachedelic** [comment]
Would you be happy if the setting was defined in `strings.xml`, like [these](https://github.com/capacitor-community/background-geolocation#android)?

**Comment by medvidekPanda** [comment]
I'm afraid not :-(. The requirement is to change the timeout from FE. So it can be up to 15 seconds or 60 seconds depending on the user.

**Comment by diachedelic** [comment]
There are three methods on `LocationRequest` which you have mentioned:

- setInterval
- setFastestInterval
- setMaxWaitTime

Are you looking to control every one of these?

There is also the option of pulling the Android source code into your own project. It is just [two classes](https://github.com/capacitor-community/background-geolocation/tree/master/android/src/main/java/com/equimaps/capacitor_background_geolocation) and a [few imports](https://github.com/capacitor-community/background-geolocation/blob/master/android/build.gradle). Then you can have as much control as you need.

**Comment by gine** [comment]
I would like a feature like this for this reason:

On my application i set distance filter to 5 meter and normally i receive a location with a speed value. 

I also use this speed value to do things(example: more speed value = more volume, but it can be 'distribute more salt on the road' so if speed is 0, i must distribute 0 salt and the vehicle run at 10km/h i must distribute 1kg salt, at 20km/h 2kg salt ...)

But what happens if in less then 5 meter the vehicle stops to run? I doesn't receive any update and i continue to think that my car is in motion and i continue to distribute salt. A mount of salt :)

At this point i must control myself that the last update is old and reset my speed.

**Comment by Callan003** [comment]
Hello,

This plugin works perfectly fine on both iOS and Android. However it drains battery very fast. My app requires user to have it's location tracked all the time (well almost all the time ... it's a delivery app).

I tried using stale position true and distance filter of 50m, but it had no impact on the battery usage.

My suggestion is to have the option to have the location checked every lets say 30 sec, then the service will go 'dormant' and not check for device location for the next 30 seconds. Might be stupid but I can't think of any other way I can save the battery usage.

Thank-you!

**Comment by diachedelic** [comment]
> I tried using stale position true and distance filter of 50m, but it had no impact on the battery usage.

That's a shame the distance filter does not help with battery drain. Was this the case on both iOS and Android? I think I have to accept that this plugin needs a setting which controls Android's `setInterval`. But I don't know how to help you with iOS.

On Android, it is also possible to use the [PRIORITY_BALANCED_POWER_ACCURACY](https://developers.google.com/android/reference/com/google/android/gms/location/LocationRequest#constants) priority for coarser location updates, about 100m. I would be interested to know how using this priority [instead](https://github.com/capacitor-community/background-geolocation/blob/a74c4c50a81c86b58ee271cc9a0d81b30237d9c0/android/src/main/java/com/equimaps/capacitor_background_geolocation/BackgroundGeolocationService.java#L76) affects battery usage on your device. Could you please see if that gives an improvement? It should be possible to just modify the file and rebuild.

**Comment by gine** [comment]
> @gine, I was intrigued by your idea in #41 to set the distance filter dynamically. I do not know if it would reduce energy consumption, but it might. What do you think of [this](https://gist.github.com/diachedelic/84bf9bb542c2a540b086f4af871cf891) as a solution?

it can work.

**Comment by josuelmm** [comment]
Any news on this matter?

**Comment by diachedelic** [comment]
@josuelmm What is your use case?

**Comment by josuelmm** [comment]
Being able to change the values of:
setMaxWaitTime(1000);
setInterval(1000);

From Ionic

**Comment by Zack-Heisnberg** [comment]
Hello Sir , i am not sure if you still need fix , i did the exact same thing as i needed the same thing as you xD


so , what you need to do is simple

in 
BackgroundGeolocationService.java
Change
 locationRequest.setInterval(1000); to   locationRequest.setInterval(30000);
 
 and set config distanceFilter: 0
 
 you will receive the location every 30 seconds even if you don't move :)

**Comment by matudelatower** [comment]
> Hello Sir , i am not sure if you still need fix , i did the exact same thing as i needed the same thing as you xD
> 
> so , what you need to do is simple
> 
> in BackgroundGeolocationService.java Change locationRequest.setInterval(1000); to locationRequest.setInterval(30000);
> 
> and set config distanceFilter: 0
> 
> you will receive the location every 30 seconds even if you don't move :)
> 
> , i made pull request with editable interval from capacitor
> 
> https://github.com/Zack-Heisnberg/background-geolocation

Can you make a PR?

**Comment by indrapalijama** [comment]
> Hello Sir , i am not sure if you still need fix , i did the exact same thing as i needed the same thing as you xD
> 
> so , what you need to do is simple
> 
> in BackgroundGeolocationService.java Change locationRequest.setInterval(1000); to locationRequest.setInterval(30000);
> 
> and set config distanceFilter: 0
> 
> you will receive the location every 30 seconds even if you don't move :)
> 
> , i made pull request with editable interval from capacitor
> 
> https://github.com/Zack-Heisnberg/background-geolocation

hi! is this fix applicable on android only?

**Comment by indrapalijama** [comment]
> > > Hello Sir , i am not sure if you still need fix , i did the exact same thing as i needed the same thing as you xD
> > > so , what you need to do is simple
> > > in BackgroundGeolocationService.java Change locationRequest.setInterval(1000); to locationRequest.setInterval(30000);
> > > and set config distanceFilter: 0
> > > you will receive the location every 30 seconds even if you don't move :)
> > > , i made pull request with editable interval from capacitor
> > > https://github.com/Zack-Heisnberg/background-geolocation
> > 
> > 
> > hi! is this fix applicable on android only?
> 
> yes, its only for android

i see, noob question : how can i install zack's forked repo? i tried using npm install <url> but got failed

**Comment by diachedelic** [comment]
Did you try `npm install https://github.com/Zack-Heisnberg/background-geolocation`? It worked for me.

**Comment by indrapalijama** [comment]
> npm install https://github.com/Zack-Heisnberg/background-geolocation

still got the error dear sir, here's the [error](https://pasteboard.co/h02I4HB8wmqP.png)

**Comment by diachedelic** [comment]
Did you first remove the @capacity-community/background-geolocation package from your package.json? It can not coexist with Zack's fork.

**Comment by indrapalijama** [comment]
> Did you first remove the @capacity-community/background-geolocation package from your package.json? It can not coexist with Zack's fork.

yes, i'm well aware of it, here's the [proof](https://pasteboard.co/NjpRACTAxjS4.png)

**Comment by DocBO** [comment]
If you stumble across this issue also, consider to change the relevant lines in your local folder node_modules/..../BackgroundGelolocationService.java before building again according to the new android API:

LocationRequest.Builder locationRequestBuilder = new LocationRequest.Builder(30000);
locationRequestBuilder.setPriority(Priority.PRIORITY_BALANCED_POWER_ACCURACY);
locationRequestBuilder.setIntervalMillis(30000);
locationRequestBuilder.setMaxUpdateDelayMillis(30000);
locationRequestBuilder.setMinUpdateDistanceMeters(distanceFilter);

This seems to result in a reasonable background action (I have not tested the battery drain over time).

**Comment by 1MarcoP** [comment]
Hi, I just stumbled across this plugin which works really well. 

For my use case, I want to periodically check the location of some users while they approach a certain destination and sync it (if the location changed) with the backend for some dispatcher to be informed. I want this to be as energy efficient as possible, as I often only have to request the location every few minutes. 

So far I read that the iOS part is making difficulties regarding the timeout parameter.

I have little knowledge of iOS and swift, but just tried out if a native timer can be used. 

````swift
if let interval = call.getInt("interval") {
  manager.startUpdatingLocation()
  Timer.scheduledTimer(withTimeInterval: TimeInterval(interval), repeats: true) { _ in
    manager.requestLocation()
  }
}
````

However, inside `locationManager/didUpdateLocations` I had to call `locationManager.stopUpdatingLocation()`. I thought this might be a big problem as the app won't send updates while in background, but in my tests it continued requesting the location every `n` seconds and the native timer did work surprisingly well. 

As I stated, I don't have much knowledge about swift, but would like to understand what the biggest concern is? 
Would the native timer actually be more energy efficient?

**Comment by diachedelic** [comment]
I'm surprised that works in the background. What interval values did you try? Did the status bar turn blue whilst the app was in the background?

If this approach works in the background, turns out to be energy efficient, and gives accurate results, I will add an "interval" option. However,

**Comment by 1MarcoP** [comment]
Hi, sorry for the late reply. I didn't have access to a Mac for building iOS, during the last days. 

I tried 10 seconds, 30 seconds, 1 minute, 5 minutes and 15 minutes. 
I let the 15 minute one run for about 2h 30min in the background.

However, the status bar turned blue in the beginning, but afterwards it only turned blue again when the GPS tracking was done (every `n` seconds). Therefore, I am not sure if the app could be killed in the background when the phone is under a heavy load?

**Comment by GeekAliAkbar** [comment]
After adding this 

> If you stumble across this issue also, consider to change the relevant lines in your local folder node_modules/..../BackgroundGelolocationService.java before building again according to the new android API:
> 
> LocationRequest.Builder locationRequestBuilder = new LocationRequest.Builder(30000); locationRequestBuilder.setPriority(Priority.PRIORITY_BALANCED_POWER_ACCURACY); locationRequestBuilder.setIntervalMillis(30000); locationRequestBuilder.setMaxUpdateDelayMillis(30000); locationRequestBuilder.setMinUpdateDistanceMeters(distanceFilter);
> 
> This seems to result in a reasonable background action (I have not tested the battery drain over time).

I tried above solution and got this error 
`watcher.locationRequest = locationRequest;
                                      ^
  symbol:   variable locationRequest
  location: class BackgroundGeolocationService.LocalBinder`

Any idea how to solve this

**Comment by DocBO** [comment]
Hi, I forgot this line at the end (again, according to the new android api):

LocationRequest locationRequest = locationRequestBuilder.build();

**Comment by alindzon** [comment]
as long as you are adding parameters i also think you should add numUpdates with a default of 0.  This line of code uses that parameter.

if (numUpdates>0) locationRequest.setNumUpdates(numUpdates);// used to stop after numUpdates

**Comment by diachedelic** [comment]
If we're allowing some LocationRequest.Builder methods to be configured we might as well support all of them.

I will consider a pull request that implements all of the methods listed at https://developers.google.com/android/reference/com/google/android/gms/location/LocationRequest.Builder#public-methods. Option names should be prefixed with "android", e.g. `androidGranularity`. Options corresponding to methods that expect an enum should be passed as integers, e.g. `androidGranularity: 2`. TypeScript definitions and documentation must be updated with the new options.
